content/2017-09-19-python-on-os-x.md

```diff
@@ -0,0 +1,204 @@
+you wanted to &mdash; not even with `sudo` &mdash; it's [restricted][8].
+
+Instead, most choose one of
+
+* the [pyenv][3] version manager
+* a custom Python from [Homebrew][4]
+* the [conda][7] package management system
+* "official" Python directly from [Python.org][6]
+
+### What To Do
+
+I **strongly** recommend using `pyenv`. Once installed and configured it
+allows easy switching between versions and more importantly allows you
+to have all versions owned by you (not `root`) and in one easy to find
+(and easy to remove) place.
+
+For example, the system Python is `2.7.10` and was compiled with LLVM 7.0.0:
+
+```
+$ /System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7 -c \
+>   'import sys; print(sys.version)'
+2.7.10 (default, Oct 23 2015, 19:19:21)
+[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]
+```
+
+However, with `pyenv` installed, we can use the latest[ref]2.7.14 came out
+only a few days before this post[/ref] version of 2.7[ref]I would like
+to [encourage][10] you to use Python 3, and am just using 2.7 here
+because that is the system Python[/ref]:
+
+```
+$ python2.7
+Python 2.7.14 (default, Sep 19 2017, 20:32:14)
+[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)] on darwin
+Type "help", "copyright", "credits" or "license" for more information.
+>>> import sys
+>>> sys.executable
+'${HOME}/.pyenv/versions/2.7.14/bin/python'
+```
+
+Multiple Python executables live peacefully together:
+
+```
+$ python -c 'import sys; print(sys.executable)'
+${HOME}/.pyenv/versions/3.6.2/bin/python
+$ python3.5 -c 'import sys; print(sys.executable)'
+${HOME}/.pyenv/versions/3.5.4/bin/python3.5
+```
+
+When compared with the ease of `pyenv install ${VERSION}` without any root
+privileges, downloading a `.pkg` from Python.org and running as administrator
+seems unnecessary. However, I recently found a reason why the "official"
+OS X Python binaries are worth having.
+
+### Spinning Wheels
+
+I recently decided I should be a "good citizen" and provide built Python
+wheels for a [package][5] of mine. Along the process, I came across
+[Matthew Brett][2]'s wonderful [Spinning Wheels][1].
+
+TL;DR: it's best to build wheels that target OS X with the Python
+executables from Python.org.
+
+The primary difference between a `pyenv` installed Python and an official
+one can be seen in the platform tag:
+
+```
+$ ${HOME}/.pyenv/versions/3.6.2/bin/python -c \
+>   'import distutils.util; print(distutils.util.get_platform())'
+macosx-10.11-x86_64
+$ /Library/Frameworks/Python.framework/Versions/3.6/bin/python3 -c \
+>   'import distutils.util; print(distutils.util.get_platform())'
+macosx-10.6-intel
+```
+
+First, the Python.org Python[ref]the one that starts with `/Library/...`[/ref]
+supports an older SDK (10.6) which gives **forward compatibility** of any
+binaries produced for any SDK after 10.6.
+
+Second, the `intel` architecture means the executable is a [universal][11]
```

Comment:
What does this mean for your binary extension? Does your binary extension also need to be a universal darwin shared object?

Reply:
It will be "automatically" if the default compiler flags from `distutils` are used with C and C++ sources. For other sources not explicitly supported by `distutils` I had to do a bit more jumping through hoops ([here][1] and [here][2]).

I don't think it's worth going in depth on wheels though?

I also meant to ask: do you know if there is any existing source for the "don't use system Python" argument? I feel like I've seen a blog post or talk from @glyph but I couldn't find one when Googling around.

[1]: https://github.com/matthew-brett/delocate/issues/23
[2]: https://github.com/scipy/scipy/issues/7564

Reply:
The python guide mentions it (softly) here: http://docs.python-guide.org/en/latest/starting/installation/

Our own guide mentions it here: https://cloud.google.com/python/setup#mac_os_x